### ISSUE TITLE
Error: Connection refused between two containers

### ISSUE THREAD

**Issue by Zorin95670** [testcontainers/testcontainers-java#4665]
Hi,

I have 1.16.2 version of TestContainers.

I try to create two containers one postgresql and one tomcat with war in it.

I succeed to create both.

The problem is, that the tomcat application doesn't happen to connect on the database. But when I'm trying to connect with psql client it work.

this is my code: 
```java
Network network = Network.newNetwork()
PostgreSQLContainer postgresqlContainer = (PostgreSQLContainer) new PostgreSQLContainer("postgres:13.3")
                .withInitScript("db/init.sql")
                .withDatabaseName("database_name")
                .withUsername("user_name")
                .withPassword("password")
                .withNetwork(network);
postgresqlContainer.start();

ImageFromDockerfile dockerfile = new ImageFromDockerfile().withDockerfile(new File(System.getProperty("user.dir"), "Dockerfile").toPath());
GenericContainer container = new GenericContainer(dockerfile)
                .withExposedPorts(8080)
                .withNetwork(network)
                .withLogConsumer(new Slf4jLogConsumer(LOGGER))
                .waitingFor(Wait.forHttp("Endpoint to get version of application"));
container.withEnv("database.jdbc", postgresqlContainer.getJdbcUrl());
container.start();
```

And I have this error in the tomcat container:
```
2021-11-08 10:56:12,172 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: Message    : Connection to localhost:49291 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
2021-11-08 10:56:12,172 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 
2021-11-08 10:56:12,174 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.flywaydb.core.internal.jdbc.JdbcUtils.openConnection(JdbcUtils.java:67)
2021-11-08 10:56:12,175 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.flywaydb.core.internal.jdbc.JdbcConnectionFactory.<init>(JdbcConnectionFactory.java:68)
2021-11-08 10:56:12,175 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.flywaydb.core.Flyway.execute(Flyway.java:506)
2021-11-08 10:56:12,175 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.flywaydb.core.Flyway.migrate(Flyway.java:172)
2021-11-08 10:56:12,176 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-11-08 10:56:12,176 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
2021-11-08 10:56:12,176 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-11-08 10:56:12,176 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
2021-11-08 10:56:12,176 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at or
2021-11-08 10:56:12,176 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: g.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeCustomInitMethod(AbstractAutowireCapableBeanFactory.java:1912)
2021-11-08 10:56:12,176 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1854)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1782)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	... 53 common frames omitted
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: Caused by: org.postgresql.util.PSQLException: Connection to localhost:49291 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:303)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:51)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.Driver.makeConnection(Driver.java:465)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.Driver.connect(Driver.java:264)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:677)
2021-11-08 10:56:12,177 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:189)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.springframework.jdbc.datasource.DriverManagerDataSource.getConnectionFromDriverManager(DriverManagerDataSource.java:155)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.springframework.jdbc.datasource.DriverManagerDataSource.getConnectionFromDriver(DriverManagerDataSource.java:146)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.springframework.jdbc.datasource.AbstractDriverBasedDataSource.getConnectionFromDriver(AbstractDriverBasedDataSource.java:205)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.springframework.jdbc.datasource.AbstractDriverBasedDataSource.getConnection(AbstractDriverBasedDataSource.java:169)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.flywaydb.core.internal.jdbc.JdbcUtils.openConnection(JdbcUtils.java:51)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	... 63 common frames omitted
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: Caused by: java.net.ConnectException: Connection refused
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:669)
2021-11-08 10:56:12,178 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:333)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at java.base/java.net.Socket.connect(Socket.java:645)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.core.PGStream.createSocket(PGStream.java:231)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.core.PGStream.<init>(PGStream.java:95)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:98)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:213)
2021-11-08 10:56:12,179 [docker-java-stream--2054753192] INFO  [] com.octo.cucumber.StepDefinitions - STDOUT: 	... 74 common frames omitted
```

And there is my Dockerfile:
```
FROM maven:3.8.1-openjdk-16 as builder
RUN mkdir -p /root/.m2 && mkdir /root/.m2/repository
COPY . /app/
RUN cd /app && mvn clean package -Dmaven.test.skip=true

FROM tomcat:9.0-jdk16-openjdk
COPY --from=builder /app/target/octo-spy.war /usr/local/tomcat/webapps
ENV CATALINA_OPTS=""
EXPOSE 8080
CMD ["catalina.sh", "run"]
```

**Comment by Zorin95670** [comment]
And I used this dependency for TestContainer:
```
        <dependency>
            <groupId>org.testcontainers</groupId>
            <artifactId>testcontainers</artifactId>
            <version>1.16.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.testcontainers</groupId>
            <artifactId>postgresql</artifactId>
            <version>1.16.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.testcontainers</groupId>
            <artifactId>junit-jupiter</artifactId>
            <version>1.16.2</version>
            <scope>test</scope>
        </dependency>
        ```

**Comment by bsideup** [comment]
Hi @Zorin95670,

Please see https://www.testcontainers.org/features/networking/#advanced-networking

tl;dr: you shouldn't be passing `getJdbcUrl()`'s value as `database.jdbc` to a container in the same network (as this is the value for your tests host), but instead use database's alias and the original port.

**Comment by Zorin95670** [comment]
I try this and it work, thank you :)
